replica/database.cc

```diff
@@ -1027,12 +1027,10 @@ future<> database::add_column_family(keyspace& ks, schema_ptr schema, column_fam
     ks.add_or_update_column_family(schema);
     cf->start();
     schema->registry_entry()->set_table(cf->weak_from_this());
-    _tables_metadata._column_families.emplace(uuid, std::move(cf));
-    _tables_metadata._ks_cf_to_uuid.emplace(std::move(kscf), uuid);
+    co_await _tables_metadata.add_table(schema);
```

Comment:
@Deexie This yield introduces a short period of time where schema registry is out of sync with the column families map. Causing hard to debug races.

https://github.com/scylladb/scylladb/issues/19833#issuecomment-2258308552


Guys please stop introducing yields everywhere without being super duper extra careful. I have spent countless hours debugging concurrency bugs due to these "harmless" yields. In gossiper, when updating erm, now when updating schema.

Reply:
You're right, but the solution isn't to stop adding yields, we need to have a more yield-tolerant design.

Where a path isn't yield tolerant for performance reasons, we should have a guard to assert it and crash and debug/dev modes.

Reply:
> You're right, but the solution isn't to stop adding yields, we need to have a more yield-tolerant design.

Some data structures just need to be updated atomically.
And accessing them cannot be protected by a lock (or at least we should avoid it) because they lie on a hot path -- like the schema registry and table map.

In such cases I think it's completely reasonable to require updating such data structures without a yield.

> Where a path isn't yield tolerant for performance reasons, we should have a guard to assert it and crash and debug/dev modes.

How would such guard look like though? Every separate case requires reasoning specific to that case.